### PR TITLE
BridgeMethodHandlerFactory: use MethodHandle in preference to legacy reflection

### LINKED ISSUE #1188: jdbi3 InvocationTargetException instead of UnableToExecuteStatementException on PostgreSQL
This started after bumping dropwizard from 1.3.2 to 1.3.4. That means jdbi3 from 3.1.0 to 3.2.1. I have also tried 3.3.0 and the result seems to be the same.

We have an onDemand jdbi object and calling any method that returns a legitimate SQL error now throws `InvocationTargetException `instead of the old `UnableToExecuteStatementException` which we handle to return a proper error message.

```
Jdbi jdbi = new JdbiFactory().build(environment, configuration.getDatabase(), "db"); // factory from dropwizard
jdbi.installPlugin(new PostgresPlugin());
IdentityProviderDao dao = jdbi.onDemand(IdentityProviderDao.class)
dao.insert(entity);
```

Exception sample:
```
! org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "identity_provider_display_name_key"
!   Detail: Key (display_name)=(cz_test_ldap) already exists.
! at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2433)
! at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2178)
! at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:306)
! at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
! at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
! at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:155)
! at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:144)
! at sun.reflect.GeneratedMethodAccessor263.invoke(Unknown Source)
! at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
! at java.lang.reflect.Method.invoke(Method.java:498)
! at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:114)
! at com.sun.proxy.$Proxy66.execute(Unknown Source)
! at org.jdbi.v3.core.statement.SqlLoggerUtil.wrap(SqlLoggerUtil.java:29)
! at org.jdbi.v3.core.statement.SqlStatement.internalExecute(SqlStatement.java:1380)
! ... 111 common frames omitted
! Causing: org.jdbi.v3.core.statement.UnableToExecuteStatementException: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "identity_provider_display_name_key"
!   Detail: Key (display_name)=(cz_test_ldap) already exists. [statement: cropped', parameters=ParsedParameters{positional=false, parameterNames=[cropped]}}", arguments:{ positional:{}, named:{}, finder:[{lazy bean property arguments "cropped"]}]
! at org.jdbi.v3.core.statement.SqlStatement.internalExecute(SqlStatement.java:1388)
! at org.jdbi.v3.core.result.ResultProducers.lambda$returningUpdateCount$0(ResultProducers.java:48)
! at org.jdbi.v3.core.statement.Update.execute(Update.java:53)
! at org.jdbi.v3.core.statement.Update.execute(Update.java:41)
! at org.jdbi.v3.sqlobject.statement.internal.SqlUpdateHandler.lambda$new$1(SqlUpdateHandler.java:59)
! at org.jdbi.v3.sqlobject.statement.internal.SqlUpdateHandler.lambda$configureReturner$3(SqlUpdateHandler.java:74)
! at org.jdbi.v3.sqlobject.statement.internal.CustomizingStatementHandler.invoke(CustomizingStatementHandler.java:155)
! at org.jdbi.v3.sqlobject.statement.internal.SqlUpdateHandler.invoke(SqlUpdateHandler.java:30)
! at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$null$13(SqlObjectFactory.java:158)
! at org.jdbi.v3.core.internal.JdbiThreadLocals.invokeInContext(JdbiThreadLocals.java:27)
! at org.jdbi.v3.core.LazyHandleSupplier.lambda$invokeInContext$1(LazyHandleSupplier.java:68)
! at org.jdbi.v3.core.internal.JdbiThreadLocals.invokeInContext(JdbiThreadLocals.java:27)
! at org.jdbi.v3.core.LazyHandleSupplier.invokeInContext(LazyHandleSupplier.java:67)
! at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$createInvocationHandler$14(SqlObjectFactory.java:157)
! at com.sun.proxy.$Proxy79.insertIdentityProvider(Unknown Source)
! at com.cyxtera.controller.service.identity.identityprovider.dao.IdentityProviderDao.insert(IdentityProviderDao.java:103)
! at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
! at org.jdbi.v3.sqlobject.DefaultMethodHandler.invoke(DefaultMethodHandler.java:102)
! at org.jdbi.v3.sqlobject.transaction.internal.TransactionDecorator.lambda$decorateHandler$1(TransactionDecorator.java:47)
! at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$null$13(SqlObjectFactory.java:158)
! at org.jdbi.v3.core.internal.JdbiThreadLocals.invokeInContext(JdbiThreadLocals.java:27)
! at org.jdbi.v3.core.LazyHandleSupplier.lambda$invokeInContext$1(LazyHandleSupplier.java:68)
! at org.jdbi.v3.core.internal.JdbiThreadLocals.invokeInContext(JdbiThreadLocals.java:27)
! at org.jdbi.v3.core.LazyHandleSupplier.invokeInContext(LazyHandleSupplier.java:67)
! at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$createInvocationHandler$14(SqlObjectFactory.java:157)
! at com.sun.proxy.$Proxy79.insert(Unknown Source)
! ... 86 common frames omitted
! Causing: java.lang.reflect.InvocationTargetException: null
! at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
! at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
! at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
! at java.lang.reflect.Method.invoke(Method.java:498)
! at org.jdbi.v3.sqlobject.BridgeMethodHandlerFactory.lambda$null$3(BridgeMethodHandlerFactory.java:39)
! at org.jdbi.v3.sqlobject.transaction.internal.TransactionDecorator.lambda$null$0(TransactionDecorator.java:50)
! at org.jdbi.v3.core.transaction.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:135)
! at org.jdbi.v3.core.Handle.inTransaction(Handle.java:393)
! at org.jdbi.v3.sqlobject.transaction.internal.TransactionDecorator.lambda$decorateHandler$1(TransactionDecorator.java:59)
! at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$null$13(SqlObjectFactory.java:158)
! at org.jdbi.v3.core.internal.JdbiThreadLocals.invokeInContext(JdbiThreadLocals.java:27)
! at org.jdbi.v3.core.LazyHandleSupplier.lambda$invokeInContext$1(LazyHandleSupplier.java:68)
! at org.jdbi.v3.core.internal.JdbiThreadLocals.invokeInContext(JdbiThreadLocals.java:27)
! at org.jdbi.v3.core.LazyHandleSupplier.invokeInContext(LazyHandleSupplier.java:67)
! at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$createInvocationHandler$14(SqlObjectFactory.java:157)
! at org.jdbi.v3.core.OnDemandExtensions.invoke(OnDemandExtensions.java:76)
! ... 70 common frames omitted
! Causing: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
! at org.jdbi.v3.core.OnDemandExtensions.invoke(OnDemandExtensions.java:82)
! at org.jdbi.v3.core.OnDemandExtensions.lambda$null$0(OnDemandExtensions.java:63)
! at org.jdbi.v3.core.internal.JdbiThreadLocals.invokeInContext(JdbiThreadLocals.java:27)
! at org.jdbi.v3.core.OnDemandExtensions.lambda$null$1(OnDemandExtensions.java:62)
! at org.jdbi.v3.core.Jdbi.withExtension(Jdbi.java:439)
! at org.jdbi.v3.core.OnDemandExtensions.lambda$create$2(OnDemandExtensions.java:61)
! at com.sun.proxy.$Proxy79.insert(Unknown Source)
! at com.cyxtera.controller.dropwizard.BaseEntityResource.save(BaseEntityResource.java:269)
! at com.cyxtera.controller.service.identity.resource.IdentityProviderResource.save(IdentityProviderResource.java:146)
! at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
! at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
! at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
! at java.lang.reflect.Method.invoke(Method.java:498)
! at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
! at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)
! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
! at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
! at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
! at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
! at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
! at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
! at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
! at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
! at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
! at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
! at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
! at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
! at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
! at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
! at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
! at io.dropwizard.jetty.NonblockingServletHolder.handle(NonblockingServletHolder.java:49)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
! at io.dropwizard.servlets.ThreadNameFilter.doFilter(ThreadNameFilter.java:35)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
! at io.dropwizard.jersey.filter.AllowedMethodsFilter.handle(AllowedMethodsFilter.java:45)
! at io.dropwizard.jersey.filter.AllowedMethodsFilter.doFilter(AllowedMethodsFilter.java:39)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
! at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
! at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
! at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
! at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
! at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
! at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:239)
! at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
! at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:674)
! at io.dropwizard.jetty.BiDiGzipHandler.handle(BiDiGzipHandler.java:67)
! at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:56)
! at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:169)
! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
! at org.eclipse.jetty.server.Server.handle(Server.java:531)
! at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
! at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
! at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
! at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
! at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:132)
! at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:760)
! at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:678)
! at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by stevenschlansker**
This prevents any thrown exceptions from getting unexpectedly wrapped in InvocationTargetException

Fixes #1188

### CODE REVIEW

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java [hunk 1]
```diff
@@ -36,7 +37,30 @@
                 return IntStream.range(0, method.getParameterCount())
                     .allMatch(i -> methodParamTypes[i].isAssignableFrom(candidateParamTypes[i]));
             })
-            .<Handler>map(m -> (target, args, handle) -> m.invoke(target, args))
+            .<Handler>map(m -> {
+                final MethodHandle mh = unreflect(sqlObjectType, m);
```
Comment (arteam) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java]: I think we could move `MethodHandle` inside the invoking lambda and just pass it around:
```java
(target, args, handle) -> {
    try {
        return unreflect(sqlObjectType, m).bindTo(target).invokeWithArguments(args);
    } catch (Throwable t) { // Handle et al should <X extends Throwable> not Exception!!!
         sneakyThrow(t);
         throw new AssertionError("unreachable", t);
    }
}
```

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java [hunk 2]
```diff
@@ -36,7 +37,30 @@
...
+                return (target, args, handle) -> {
+                    try {
+                        return mh.bindTo(target).invokeWithArguments(args);
+                    } catch (Throwable t) { // Handle et al should <X extends Throwable> not Exception!!!
+                        sneakyThrow(t);
```
Comment (arteam) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java]: The Java compiler is horrible here :(

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java (hunk 1 above)
Comment (stevenschlansker) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java]: I pulled it up so as to cache the lookup once up front.  I don't know how expensive it is, but it doesn't vary call to call...

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java (hunk 2 above)
Comment (stevenschlansker) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java]: Yes.  Also, I think it shows that we erred in our `<X extends Exception>` it should have been `<X extends Throwable>`

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBridgeException.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    public void setUp() {
+        final Handle h = this.dbRule.getSharedHandle();
+        h.execute("CREATE TABLE uniq (id INTEGER PRIMARY KEY)");
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    @Test
+    public void testBridgeExceptionPassthru() throws Throwable {
+        final ExceptionalBridge dao = this.dbRule.getSharedHandle().attach(ExceptionallyTypedBridge.class);
+        final Object arg = 3;
```
[the remark this answers is not in the document]
Comment (stevenschlansker) [reply sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBridgeException.java]: The whole point is to call through the bridge method.  If you have a static type of `Integer`, you no longer call through the bridge method, and no longer trigger the bug.

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBridgeException.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import org.junit.Test;
+
+public class TestBridgeException {
+
+    @Rule
+    public H2DatabaseRule dbRule = new H2DatabaseRule().withPlugin(new SqlObjectPlugin());
+
+    @Before
+    public void setUp() {
+        final Handle h = this.dbRule.getSharedHandle();
```
[the remark this answers is not in the document]
Comment (stevenschlansker) [reply sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBridgeException.java]: copypasta :)

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java (hunk 2 above)
Comment (qualidafial) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/BridgeMethodHandlerFactory.java]: Is it time to start keeping track of these mistakes in a 4.x wishlist document, the way we did from 2.x to 3.x?

**Pull request merged**